### PR TITLE
feat(annotation-thread): handle create reply

### PR DESCRIPTION
**Pull request by jankowiakdawid**
New component `ActivityThreadReplyForm`
  * shows Reply button
  * handles showing/hiding `CommentForm`

Changes in `ActivityThread`
  * renders `ActivityThreadReplyForm`

Changes in `CommentForm.js`
  * adds new optional prop `placeholder`

#### Designs
From figma (don't mind the background it's part of different feature, and the black blocks on right are part of arrows showing the flow)
Screenshot 2022-09-30 at 14 03 05
Screenshot 2022-09-30 at 14 03 38

As it's implemented
Screenshot 2022-09-30 at 14 04 59
Screenshot 2022-09-30 at 14 05 16

### CODE REVIEW

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import messages from './messages';
+import './ActivityThreadReplyForm.scss';
+
+type Props = {
+    currentUser?: User,
+    getMentionWithQuery?: Function,
+    getUserProfileUrl?: GetProfileUrlCallback,
+    mentionSelectorContacts?: SelectorItems<>,
+    onReplyCreate: Function,
```
Comment (kkuliczkowski-box) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:22]: Please change the type here to arrow function that specifies expected attributes and their types (the same way it's defined in `ActivityThread`). This also applies to `getMentionWithQuery` above.

File: src/elements/content-sidebar/activity-feed/activity-feed/__tests__/ActivityThreadRepliesForm.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+// @flow
+import React from 'react';
+import { render, screen, fireEvent } from '@testing-library/react';
+import { IntlProvider } from 'react-intl';
+import localize from '../../../../../../test/support/i18n';
+import messages from '../messages';
+import ActivityThreadReplyForm from '../ActivityThreadReplyForm.js';
+
+const noop = () => {};
```
Comment (kkuliczkowski-box) [inline src/elements/content-sidebar/activity-feed/activity-feed/__tests__/ActivityThreadRepliesForm.test.js:9]: We typically use `noop` from lodash.

File: src/elements/content-sidebar/activity-feed/comment-form/CommentForm.js [hunk 3]
```diff
@@ -149,7 +151,9 @@ class CommentForm extends React.Component<Props, State> {
                             onChange={this.onMentionSelectorChangeHandler}
                             onFocus={onFocus}
                             onMention={getMentionWithQuery}
-                            placeholder={tagged_message ? undefined : formatMessage(messages.commentWrite)}
+                            placeholder={
+                                tagged_message ? undefined : placeholder ?? formatMessage(messages.commentWrite)
```
Comment (kkuliczkowski-box) [inline src/elements/content-sidebar/activity-feed/comment-form/CommentForm.js:155]: I think we should have a test for this logic.

Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/comment-form/CommentForm.js:155]: Agree! I'll add it tomorrow morning

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
+// @flow
+
+import * as React from 'react';
+import { FormattedMessage, useIntl } from 'react-intl';
```
Comment (jankowiakdawid) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:4]: This currently breaks the `yarn flow check`, I see I need to extend the https://github.com/jankowiakdawid/box-ui-elements/blob/annotationThread-handleCreateReply/flow-typed/npm/react-intl_v2.x.x.js#L1 to fix it

Comment (greg-in-a-box) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:4]: There is a script that checks for `useIntl` being used, I am unsure why exactly but from what I understand we should be using the `intl` from via `injectIntl` to formatMessage e.g. https://github.com/box/box-ui-elements/blob/830bf71bef5720a5e6ba128afa35152112674c17/src/elements/content-explorer/ShareDialog.js#L54

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js [hunk 5]
```diff
@@ -13,6 +13,8 @@ import type { SelectorItems, User } from '../../../../common/types/core';
 import type { BoxCommentPermission, Comment as CommentType, FeedItemStatus } from '../../../../common/types/feed';
 
 import messages from './messages';
+import ActivityThreadReplyForm from './ActivityThreadReplyForm';
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:16]: Minor, but please move this up near `import ActivityThreadReplies`

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js [hunk 6]
```diff
@@ -106,6 +109,16 @@ const ActivityThread = ({
                     translations={translations}
                 />
             )}
+
+            {onReplyCreate ? (
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:113]: It looks like `onReplyCreate` will always be defined due to the default `noop` value being assigned above during destructuring. Is `onReplyCreate` nullable? If so, its type doesn't indicate that.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplies.js [hunk 7]
```diff
@@ -37,9 +37,7 @@ const ActivityThreadReplies = ({
     const lastReply = replies[replies.length - 1];
 
     const getReplyPermissions = (reply: CommentType): BoxCommentPermission => {
-        const {
-            permissions: { can_delete = false, can_edit = false, can_resolve = false },
-        } = reply;
+        const { permissions: { can_delete = false, can_edit = false, can_resolve = false } = {} } = reply;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplies.js:40]: In what scenario would `permissions` be undefined? Its type indicates it's a required field. If this is for testing purposes only, we should avoid making logical changes to support tests.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    const [showReplyForm, setShowReplyForm] = React.useState(false);
+    const placeholder = useIntl().formatMessage(messages.replyInThread);
+
+    return showReplyForm ? (
+        <CommentForm
+            className=""
+            isOpen
+            isEditing
+            getAvatarUrl={() => Promise.resolve()}
+            // $FlowFixMe
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:40]: Why is this needed? Can we add a brief comment explaining the need for disabling the type check here?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    mentionSelectorContacts,
+    getMentionWithQuery,
+    onReplyCreate,
+}: Props) {
+    const [showReplyForm, setShowReplyForm] = React.useState(false);
+    const placeholder = useIntl().formatMessage(messages.replyInThread);
+
+    return showReplyForm ? (
+        <CommentForm
+            className=""
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:36]: Can we omit this prop?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            createComment={({ text, hasMentions }) => {
+                onReplyCreate(text, hasMentions);
+                setShowReplyForm(false);
+            }}
+            mentionSelectorContacts={mentionSelectorContacts}
+            getMentionWithQuery={getMentionWithQuery}
+            placeholder={placeholder}
+        />
+    ) : (
+        <PlainButton role="button" className="bcs-ActivityThread-replyForm" onClick={() => setShowReplyForm(true)}>
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:52]: SUIT naming has guidelines around the top-level class name. In this case, since the component is `ActivityThreadReplyForm`, something like `bcs-ActivityThreadReplyForm-toggle` would be appropriate

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+}: Props) {
+    const [showReplyForm, setShowReplyForm] = React.useState(false);
+    const placeholder = useIntl().formatMessage(messages.replyInThread);
+
+    return showReplyForm ? (
+        <CommentForm
+            className=""
+            isOpen
+            isEditing
+            getAvatarUrl={() => Promise.resolve()}
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:39]: What is the goal of this change?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+export default function ActivityThreadReplyForm({
+    currentUser,
+    mentionSelectorContacts,
+    getMentionWithQuery,
+    onReplyCreate,
+}: Props) {
+    const [showReplyForm, setShowReplyForm] = React.useState(false);
+    const placeholder = useIntl().formatMessage(messages.replyInThread);
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:32]: We don't currently support `useIntl` for backward-compatibility reasons. Can we switch to `injectIntl`?

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.scss [hunk 13]
```diff
@@ -0,0 +1,18 @@
+@import '../../../common/variables';
+
+.bcs-ActivityThread-replyForm {
+    $replyPadding: $bdl-grid-unit $sidebarActivityFeedSpacingHorizontal ($bdl-grid-unit * 5);
+
+    display: flex;
+    align-items: center;
+    padding: $replyPadding;
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.scss:8]: Minor, but we can avoid this variable:

```
&,
&:hover {
    padding: $bdl-grid-unit $sidebarActivityFeedSpacingHorizontal ($bdl-grid-unit * 5);
}

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.scss [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+    &:hover {
+        padding: $replyPadding;
+    }
+
+    .bcs-ActivityThread-replyForm-arrow {
+        margin-right: $bdl-grid-unit;
+        transform: matrix(-1, 0, 0, 1, 0, 0);
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.scss:16]: Can we add a comment with the reasoning for this style?

File: src/elements/content-sidebar/activity-feed/comment-form/CommentForm.js [hunk 15]
```diff
@@ -119,6 +120,7 @@ class CommentForm extends React.Component<Props, State> {
             tagged_message,
             getAvatarUrl,
             showTip = true,
+            placeholder,
```
Comment (jstoffan) [inline src/elements/content-sidebar/activity-feed/comment-form/CommentForm.js:123]: Minor, but the changes here can be simplified:

```
placeholder = formatMessage(messages.commentWrite)

// ... later

placeholder={tagged_message ? undefined : placeholder}

**Comment by jstoffan** [comment]
The PR title has a typo: annotattion -> annotation

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.scss (hunk 13 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.scss:8]: Thanks! I didn't know this notation.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js (hunk 11 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:39]: As per design I don't need to show an avatar (thus `isEditing`), but in CommentForm props `getAvatarUrl` is defined as required 
https://github.com/box/box-ui-elements/blob/6b59624dcd33f95eb24193fce1683948f8fcf6c1/src/elements/content-sidebar/activity-feed/comment-form/CommentForm.js#L30
where `GetAvatarUrlCallback` is defined as
https://github.com/box/box-ui-elements/blob/6b59624dcd33f95eb24193fce1683948f8fcf6c1/src/elements/common/flowTypes.js#L25
I wanted to avoid unnecessary props for `AvtivityThreadReplyForm` and prop drilling form above.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js (hunk 9 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:36]: It appears it's required
https://github.com/box/box-ui-elements/blob/6b59624dcd33f95eb24193fce1683948f8fcf6c1/src/elements/content-sidebar/activity-feed/comment-form/CommentForm.js#L26

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js (hunk 8 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:40]: This is due to the mismatch between types
CommentForm which require `user` and the `ActivityThreadReplyForm` where it's optional.
I saw it also in `ActivityFeed`
https://github.com/box/box-ui-elements/blob/6b59624dcd33f95eb24193fce1683948f8fcf6c1/src/elements/content-sidebar/activity-feed/activity-feed/ActivityFeed.js#L368
and `Comment`
https://github.com/box/box-ui-elements/blob/6b59624dcd33f95eb24193fce1683948f8fcf6c1/src/elements/content-sidebar/activity-feed/comment/Comment.js#L239

As far as I can tell `user` is used only to show avatar, and I don't need it here (will add screenshot of designs).

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js (hunk 4 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:4]: Ok, then as @jstoffan suggested I'll use `injectIntl`

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplies.js (hunk 7 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplies.js:40]: When submitting reply, temporarily new comment is created, and displayed before the response with actual comment get back. This temporal comment doesn't have `permissions` field.
I think alternatively we can fix it by adding `permissions` here
https://github.com/box/box-ui-elements/blob/6b59624dcd33f95eb24193fce1683948f8fcf6c1/src/api/Feed.js#L1707-L1731

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js (hunk 9 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js:40]: While looking again at design I discovered that I need to change styling of `CommentForm` a bit more, so I'll use this prop for that.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js (hunk 6 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThread.js:113]: True. That was a mistake on my side. `onReplyCreate` should be optional on `ActivityTHread` but not on `ActivityThreadReplyForm`.
I'll add one more test to catch that we don't show reply button when `onReplyCreate` isn't passed.

File: src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js (hunk 1 above)
Comment (jankowiakdawid) [reply src/elements/content-sidebar/activity-feed/activity-feed/ActivityThreadReplyForm.js]: I wasn't able to find how should `getMentionWithQuery` be typed though.

File: package.json [hunk 16]
```diff
@@ -82,11 +82,11 @@
         "start:storybook": "LANGUAGE=en-US REACT=true BABEL_ENV=development NODE_ENV=development start-storybook -p 6061",
         "start:storybook:ci": "yarn start:storybook --ci",
         "start:storybook:legacy": "BROWSERSLIST_ENV=production yarn start:storybook",
-        "test": "NODE_ICU_DATA=node_modules/full-icu BABEL_ENV=test NODE_ENV=test yarn jest -c scripts/jest/jest.config.js",
+        "test": "BABEL_ENV=test NODE_ENV=test yarn jest -c scripts/jest/jest.config.js",
```
Comment (jstoffan) [inline package.json:85]: Looks like something went awry with the latest rebase.

Comment (jankowiakdawid) [reply package.json]: Yeah, should be fixed now.

**Review by jstoffan — DISMISSED** [review]
